Shapeless compilation failed on Ubuntu 24.04

When testing the Ubuntu 24.04 runner of GitHub Actions some of the compile tests failed on it:

```
  1) compile
       shapelessMean:
     AssertionError: expected [ false ] to not include false
      at <anonymous> (tests/utils.ts:15:76)
      at assertArray (tests/utils.ts:8:5)
      at assertArrayAllTrue (tests/utils.ts:15:42)
      at Context.<anonymous> (tests/compile.spec.ts:563:5)

  2) compile
       compileShapelessWithBroadcast:
     Error: [broadcast_shapes] Shapes (2,2,5) and (2,3,2,5) cannot be broadcast.
      at Context.<anonymous> (tests/compile.spec.ts:807:13)
```

I'm not sure if it is a bug in JS bindings or in upstream code, but can you consider updating this repo's CI to run on Ubuntu 24.04 to help root out some possibilities?

Indeed, I did get the same errors as you:

```
======================================================================
ERROR: test_compile_shapeless_with_broadcast (test_compile.TestCompile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/mlx/mlx/python/tests/test_compile.py", line 920, in test_compile_shapeless_with_b
    out = cfun((a, b))
ValueError: [broadcast_shapes] Shapes (2,2,5) and (2,3,2,5) cannot be broadcast.

======================================================================
FAIL: test_shapeless_mean (test_compile.TestCompile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/mlx/mlx/python/tests/test_compile.py", line 686, in test_shapeless_mean
    self.assertTrue(mx.allclose(out, mean(x)))
AssertionError: array(False, dtype=bool) is not true

----------------------------------------------------------------------
Ran 512 tests in 49.678s

FAILED (failures=1, errors=1, skipped=27)
```

I will investigate what's going on there.

--- a/mlx/transforms.cpp
+++ b/mlx/transforms.cpp
@@ -37,7 +37,7 @@ class Synchronizer : public Primitive {
 // are currently under a function transformation and the retain_graph()
 // function which returns true if we are forced to retain the graph during
 // evaluation.
-std::vector<bool> detail::InTracing::trace_stack{};
+std::vector<char> detail::InTracing::trace_stack{};
 int detail::RetainGraph::tracing_counter{0};
 
 array eval_impl(std::vector<array> outputs, bool async) {

--- a/mlx/transforms_impl.h
+++ b/mlx/transforms_impl.h
@@ -36,7 +36,7 @@ struct InTracing {
   }
 
  private:
-  static std::vector<bool> trace_stack;
+  static std::vector<char> trace_stack;
 };
 
 struct RetainGraph {

